.github/workflows/create-alpha-release-branch.yaml

```diff
@@ -0,0 +1,80 @@
+name: Create Alpha Release Branch
+
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Alpha Release Branch Version'
+        required: false
+
+env:
+  artifactId: gaffer2
+  MAVEN_OPTS: -Dmaven.wagon.http.retryHandler.count=3 -Dmaven.wagon.httpconnectionManager.ttlSeconds=25
+
+jobs:
+  create-release-branch:
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout v2-alpha
+      uses: actions/checkout@v2
+      with:
+        ref: v2-alpha
+        token: ${{ secrets.ADMIN_GITHUB_TOKEN }}
+        fetch-depth: 0
+
+    - name: Set version from input
+      if: ${{ github.event.inputs.version }}
+      run: echo "RELEASE_VERSION=$(echo ${{ github.event.inputs.version }} | sed 's/^v//')" >> $GITHUB_ENV
+
+    - name: Get latest tag
+      if: ${{ !github.event.inputs.version }}
+      uses: actions-ecosystem/action-get-latest-tag@v1
+      id: get-latest-tag
+
+    - name: Format latest tag
+      if: ${{ !github.event.inputs.version }}
+      run: echo "CURRENT_VERSION=$(echo ${{ steps.get-latest-tag.outputs.tag }} | sed 's/^gaffer2-//')" >> $GITHUB_ENV
+
+    - name: Bump latest tag variable
+      if: ${{ !github.event.inputs.version }}
+      uses: actions-ecosystem/action-bump-semver@v1
+      id: bump-semver
+      with:
+        current_version: ${{ env.CURRENT_VERSION }}
+        level: prerelease
+
+    - name: Set version automatically
+      if: ${{ !github.event.inputs.version }}
+      run: echo "RELEASE_VERSION=$(echo ${{ steps.bump-semver.outputs.new_version }} )" >> $GITHUB_ENV
+
+    - name: Verify version regex
+      uses: actions-ecosystem/action-regex-match@v2
+      id: regex-match
+      with:
+        text: ${{ env.RELEASE_VERSION }}
+        # Ensure version starts in 2
+        regex: '^(2)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
+
+    - name: Fail if incorrect version
+      if: ${{ !steps.regex-match.outputs.match || !contains(env.RELEASE_VERSION, 'alpha') }}
+      run: exit 1
+
+    - name: Set release branch
+      run: echo "BRANCH_NAME=$(echo alpha-release/${{ env.RELEASE_VERSION }} )" >> $GITHUB_ENV
+
+    - name: Set up Github credentials
+      run: |
+        git config user.name Gaffer
+        git config user.email github-actions@github.com
+
+    - name: Update versions
+      run: |
+        mvn versions:set -DnewVersion=$RELEASE_VERSION -DgenerateBackupPoms=false
+        sed -i'' -e "s/^gaffer.version=.*/gaffer.version=$RELEASE_VERSION/" rest-api/common-rest/src/main/resources/version.properties
+
+    - name: Push to release branch
+      run: |
+        git checkout -b $BRANCH_NAME
+        git commit -a -m "prepare release $artifactId-$RELEASE_VERSION"
+        git push --set-upstream origin $BRANCH_NAME
```

Comment:
Seems that a lot of the code is duplicated. If storing env variables in the in the central cache `GITHUB_ENV` keeps them for the entirety of the action (with multiple workflows) then you could have an alpha, hot-fix etc workflows which store the values such as `BRANCH_NAME` and `ARTIFACT_ID` in the cache using:
```echo "BRANCH=..." >> $GITHUB_ENV```
Then call a generic workflow where the values are retrieved from the cache.

Reply:
When you define an environment variable within a job, then that variable can only be used within that job. https://docs.github.com/en/actions/learn-github-actions/environment-variables

This is because each job is run on a different machine. The only variables that can be used across all jobs in a workflow are ones that are predefined at the top of the workflow.

To export a variable to dependent jobs you have to use outputs like this: https://github.com/gchq/Gaffer/blob/0117e3faf758d87e339bcb2513028b485f03fe1a/.github/workflows/release.yaml#L41-L46

And use them like this: https://github.com/gchq/Gaffer/blob/0117e3faf758d87e339bcb2513028b485f03fe1a/.github/workflows/release.yaml#L210

Reply:
Thats a shame...

Reply:
It may seem like duplication but state and conditionals aren't really a thing you can do easily with these workflows, otherwise it would just be a simple case of `if X: ..., else if Y: ..., else if Z: ...`
Instead my approach was to split the "conditionals" into 3 workflows and then the release workflow handles the common elements.

Reply:
Seems you can have reusable workflows with inputs:
[https://docs.github.com/en/actions/learn-github-actions/reusing-workflows](url)

Reply:
This can be done but with the create release branch action there would still be 4 conditionals with 3 distinct choices each. This means you have two choices: one action riddled with lots of ugly versions of if, else if, and else within GitHub actions syntax, or one action where the value from these conditionals are set by the user, but it seems easier for the user to just choose the correct workflow from 3 options rather than have to define the branch to merge from, name prefix of the release branch, correct regex and semver release increment every release.

Reply:
There are just a handful of variables you pass in... The files are almost identical other than the regex, branch type and one conditional

Reply:
There are 4 variables to pass in (branch to merge from, name prefix of the release branch, correct regex and semver release increment ). I think it makes the action harder to use if someone has to define those every time as options.
The other option like I said is you just input one of the 3 types (standard, hotfix or alpha) but then there has to be 4 conditionals at the top of the action that programmatically set the 4 variables to one of the 3 options which is very ugly. For example, this is how I set one variable with 2 options: https://github.com/gchq/Gaffer/blob/dd084ad45339d5b49aa1b71dd0de67a4281a724b/.github/workflows/release.yaml#L52-L58

I just think it would be ugly to have 12 of these at the start of the action and would rather have 3 different actions.

Reply:
I just had:
```
on:
  workflow_call:
    inputs:
      regex:
        required: true
        type: string
      branch:
        required: true
        type: string
        # release: "release"
        # alpha: "alpha-release"
        # hotfix: "hotfix"
      branch_type:
        required: true
        type: string
        # release: "develop"
        # alpha: "alpha-v2"
        # hotfix: "master"
      conditional_regex:
        required: false
        type: string
        # alpha: "alpha"
      version_type:
        required: true
        type: string
        # release: "minor"
        # hotfix: "patch"
        # alpha: "prerelease"
    secrets:
      token:
        required: true
        # ADMIN_GITHUB_TOKEN
```
And included extra step for the `conditional_regex`:
```
- name: Fail if doesn't contain conditional regex
      run: |
        if [[ ! -z ${{ input.conditional_regex }} && ! ${{ env.RELEASE_VERSION }} =~ ${{ input.conditional_regex }} ]] 
        then
            exit 1
        fi
```
But it's only a suggestion. Just saves having 3 near identical `create-.*.yaml` files.

Reply:
The lines that are commented out don't do anything, what happens if I submit a job with branch set to "develop"? And if you plan to let a user have to input all of the correctly manually each time that is pretty unusable. Or another workflow can trigger this, but then that workflow will be an ugly list of those conditionals that I showed above. At the end of the day you just can't do this type of thing with GitHub actions and make it pretty sadly.

Reply:
The comments were just what I would enter. In the `create-release-branch.yaml` file you would have:
```
on:
  workflow_dispatch:
    inputs:
      version:
        description: 'Release Branch Version'
        required: false

env:
  artifactId: gaffer2
  MAVEN_OPTS: -Dmaven.wagon.http.retryHandler.count=3 -Dmaven.wagon.httpconnectionManager.ttlSeconds=25

jobs:
  create-release-branch:
    uses: .github/workflows/create-branch.yml@v1
    with:
      inputs:
        regex: '<regex>'
        branch: 'release'
        branch_type: 'develop'
        version_type: 'minor'
    secrets:
      token: ${{ secrets.ADMIN_GITHUB_TOKEN }}
```
And do the same for `alpha` and `hotfix`

Reply:
So you think there should be 4 files instead of 3? I think that's adding more complexity for the sake of reducing number of lines which I don't see as an issue.

Reply:
The `create-branch.yaml` has the same code as `create-alpha-release.yaml` but with the inputs as above and values substituted for for the input variables along with the the extra step to do the regex check if a value is provided (for the other types you don't include an `conditional_regex` so it just continues).

Reply:
I suppose it would be easier to maintain but harder to understand

Reply:
Whatever you think is best. Code reuse means you don't need to alter multiple files if you make a change to the create logic...

Reply:
Yeah I think it will be easier to maintain for those that understand how it works but I worry that it is becoming increasingly complex and using more niche features of GitHub actions that are harder to understand. Maybe cleaner though, will have a think and ask some more people maybe

Reply:
So I think a clean way to reduce code duplication might just be to put it all in one file with a single input like "branch type" which can either be left blank (standard), or set to hotfix or alpha. Then there can be one big ugly block setting all the variables, then it makes the branch. This reduces files from 4 to 1 and uses less complex features.

Having said that, I think for now we should just use what I have made because it will take a long time to rejig and retest all the branch types on all 4 repos. We will have to rejig the actions in the future anyway once Gaffer 2 is released to change the alpha branch option to a Gaffer 1 support branch. If we still wish to reduce it to 1 file and 1 workflow then, it can be done. What do you think? @r32575